### PR TITLE
OpenTelemetry instrumentation for Timers

### PR DESCRIPTION
**Pull request by bachuv**
This PR adds instrumentation for Durable timers.

**Sample orchestration**
```
class HelloSequenceWithTimer : TaskOrchestration<string, string>
{
    public override async Task<string> RunTask(OrchestrationContext context, string input)
    {
        string output = "";
        output += await context.ScheduleTask<string>(typeof(SayHello), "Tokyo") + ", ";
        await context.CreateTimer(context.CurrentUtcDateTime.AddSeconds(15), true);
        output += await context.ScheduleTask<string>(typeof(SayHello), "London") + ", ";
        output += await context.ScheduleTask<string>(typeof(SayHello), "Seattle");
        return output;
    }
}
```

**Related Gantt charts**
Application Insights
image

image

Zipkin
image

image

### CODE REVIEW

**Comment by cgillum** [comment]
I'd like to get other people's opinion on this (feel free to chime in @jviau), but I think it would be more intuitive if the span for the timer lasted for the full duration, similar to activities, rather than having it cover only the few microseconds required to create the timer in memory, which the current implementation appears to be doing.

Here's what I'd expect this to look like:
image

Also, do we really need the timer history event to include the trace context? I didn't expect this to be necessary since we're not actually calling into any other code or services. Timers are just a way for an orchestration to call itself. If we could implement this in a way that didn't require trace context in the timer event, that would be ideal.

**Comment by jviau** [comment]
Yes, I agree regarding duration. Duration should be the full time this span is "active". We should also make sure to include cancelled timers and relevant attributes to identify it is canceled (maybe set it to success=false)

**Comment by bachuv** [comment]
To get the duration to reflect the time from when the timer is created to when it's fired, I was thinking of adding the trace context to the TimerFired event. That way the Activity could start when the timer is created [here in TaskOrchestrationDispatcher ](https://github.com/Azure/durabletask/blob/main/src/DurableTask.Core/TaskOrchestrationDispatcher.cs#L876) and get assigned to the TimerFiredEvent. Then, the Activity would stop when the timer is fired [here in TaskOrchestrationContext](https://github.com/Azure/durabletask/blob/main/src/DurableTask.Core/TaskOrchestrationContext.cs#L519). This involves storing the trace context in the TimerFired event though.

I'm continuing to look through the code to avoid saving the trace context to the timer history event. Any pointers on where I could look would be helpful!

**Comment by cgillum** [comment]
@bachuv I think the approach you're suggesting is mostly right, and I agree we'll need to persist the trace context in the TimerFiredEvent payload. However, stopping the trace activity in TaskOrchestrationContext.cs won't work for modern out-of-proc because that code path will never be executed. Instead, I think we'll want to add code to stop the activity trace somewhere in TaskOrchestrationDispatcher.cs (the same file we start it in, which is also nice for symmetry), which should be executed always for all supported laguages. [ReconcileMessagesWithState](https://github.com/Azure/durabletask/blob/4ef9609d43881de4bfec0f851631028ddf5ec1b9/src/DurableTask.Core/TaskOrchestrationDispatcher.cs#L656) might be a good place to do this since we're iterating through all the incoming messages/events here anyways.

**Comment by cgillum** [comment]
Actually, I want to revisit this point:

> To get the duration to reflect the time from when the timer is created to when it's fired, I was thinking of adding the trace context to the TimerFired event.

I changed my mind (again) and think we might not need to embed any trace context into TimerFiredEvent. We don't need it to get the duration. Instead, we can use the `TimerFiredEvent.Timestamp` value to get the start time of the activity span.

**Comment by bachuv** [comment]
This makes sense. We would create the `Activity` for the Timer when it's fired and wouldn't have to store any information in the history event. This seems like a clean implementation. I'll update the PR with this implementation.

**Comment by bachuv** [comment]
Looking at it again, we still need to set the parent of the Timer span which leads to me to think we still need information from TaskOrchestrationDispatcher.

**Comment by cgillum** [comment]
@bachuv I think it depends on what we want the parent to be. If the parent is a common, long-running orchestration span (similar to my [fake screenshot above](https://github.com/Azure/durabletask/pull/826#issuecomment-1317894523)) then it might not be needed. However, I've since realized that there may be problems with the idea of a long-running orchestration span. If we instead decide to go with a new orchestration span for each replay (which I think the code currently does?) then I agree we need to save some activity context state so that we can parent the timer fired event to the correct replay. I ended up needing to do this in the `durabletask-go` implementation of distributed tracing for timers.

**Comment by jviau** [comment]
> Looking at it again, we still need to set the parent of the Timer span which leads to me to think we still need information from TaskOrchestrationDispatcher.

This depends where / how this activity is started. The tracing code we have so far has a lot of explicit setting of parent context. However, parent will also automatically be populated if there is a current activity (`Activity.Current` is non-null). So, if the timer activity is started in a context where `Activity.Current` is already set to be the parent orchestration, then there is no work to be done! It will be set for us.

**Comment by jviau** [comment]
I'd like to bring up a comment I have in our spec document:

I'd like to discuss how timer spans look in a variety of scenarios. We are focusing on the "standard" scenario where orchestration starts and waits on a timer. How about other scenarios?

1. Timer is in parallel with other work. IE: orchestration starts a timer, does some other work, then eventually waits on it.
2. Timer is waited on in a `Task.WhenAny`? When the timer is the task the on yielded, vs when it is not.
3. Timer is cancelled.

I want to make sure a dedicated span is the correct choice for a timer. The alternative is an event on the orchestration running span.

**Comment by bachuv** [comment]
> @bachuv I think it depends on what we want the parent to be. If the parent is a common, long-running orchestration span (similar to my [fake screenshot above](https://github.com/Azure/durabletask/pull/826#issuecomment-1317894523)) then it might not be needed. However, I've since realized that there may be problems with the idea of a long-running orchestration span. If we instead decide to go with a new orchestration span for each replay (which I think the code currently does?) then I agree we need to save some activity context state so that we can parent the timer fired event to the correct replay. I ended up needing to do this in the `durabletask-go` implementation of distributed tracing for timers.

@cgillum We currently have the parent as the common, long-running orchestration span. [This PR](https://github.com/Azure/durabletask/pull/692) made the update from showing orchestration replays to a common span. 

You had mentioned there might be issues with having the common span like not being able to see traces for in-progress orchestrations. Should we re-instrument orchestration spans then? We should probably also keep the implementation consistent with `durabletask-go`. Do you have a PR for the changes you're making that I can take a look at?

For now, I will see if I can use `Activity.Current` as the parent (thanks @jviau for the suggestion!)

**Comment by bachuv** [comment]
> That said, I can see how they can be noisy in certain situations too, like when they are used for timeouts (often as part of scenario 2). If a timer is cancelled (scenario 3), then I could see us simply not emitting the span since it isn't actually relevant to the orchestration's execution. We'd only emit the span for completed timers. Thoughts?

I think we should only emit the span when it adds value for the customer. For scenario 2, I think it makes sense if we emit the timer span if the timer is fired. So I agree that we should emit the span for completed timers. If a timer doesn't fire then I think the customer should be able to infer why because of the related spans (like another task completed like in scenario 2) around that time, but let me know if you think otherwise.

> @bachuv I had a side chat with Jacob on this topic and I think he convinced me that it's actually better to have a single span for the entire orchestration execution, for more than just visualization purposes (for example, customers can get more useful metrics with one big span). So if that's actually already implemented, let's keep it as-is. I'll update my `durabletask-go` implementation to match this (though part of me wonders if we should do both - one big span and have another child span for each replay - but let's not worry about that for the sake of moving this PR forward).

Ok I'll leave it how it is then. I'll add a comment about adding replay spans in the spec doc.

File: src/DurableTask.Core/TaskOrchestrationContext.cs [hunk 1]
```diff
@@ -484,6 +484,12 @@ public void HandleTimerFiredEvent(TimerFiredEvent timerFiredEvent)
             {
                 LogDuplicateEvent("TimerFired", timerFiredEvent, taskId);
             }
+
+            if (!this.IsReplaying)
+            {
+                Activity activity = TraceHelper.CreateActivityForTimer(this.OrchestrationInstance, timerFiredEvent.Timestamp, timerFiredEvent.FireAt);
+                activity.Stop();
```
Comment (cgillum) [inline src/DurableTask.Core/TaskOrchestrationContext.cs:491]: This won't work for Java because we never run any code in `TaskOrchestrationContext`. The same will be true for any future language SDKs that are based on the out-of-proc V2 protocol. I suggest you instead do this in TaskOrchestrationDispatcher.cs when we're looping through all the new messages we received. This is similar to what I'm doing in `durabletask-go`.

File: src/DurableTask.Core/Tracing/TraceHelper.cs [hunk 2]
```diff
@@ -165,6 +165,24 @@ static TraceHelper()
...
+                name: "Timer",
+                kind: ActivityKind.Internal,
+                startTime: startTime,
+                parentContext: Activity.Current?.Context ?? default,
+                tags: new KeyValuePair<string, object?>[]
+                {
+                    new("dtfx.type", "timer"),
+                    new("dtfx.instance_id", instance.InstanceId),
+                    new("dtfx.execution_id", instance.ExecutionId),
+                    new("dtfx.fire_at", fireAt.ToString("o")),
```
Comment (cgillum) [inline src/DurableTask.Core/Tracing/TraceHelper.cs:180]: I suppose we can do this as part of a separate PR, but just a reminder that we'll be replacing "dtfx" with "durabletask" in these attributes.

**Review by cgillum — COMMENTED** [review]
Updated screenshots look great! Only one blocking comment from me regarding where we emit this trace activity event.

File: src/DurableTask.Core/TaskOrchestrationContext.cs (hunk 1 above)
Comment (bachuv) [reply src/DurableTask.Core/TaskOrchestrationContext.cs:491]: Right, you had mentioned that before. Sorry about that! I'll move it to TaskOrchestrationDispatcher.cs.

File: src/DurableTask.Core/History/TimerCreatedEvent.cs [hunk 3]
```diff
@@ -41,5 +43,39 @@ public TimerCreatedEvent(int eventId)
         /// </summary>
         [DataMember]
         public DateTime FireAt { get; set; }
+
+        /// <summary>
+        /// The W3C trace context associated with this event.
+        /// </summary>
+        [DataMember]
+        public DistributedTraceContext ParentTraceContext { get; set; }
```
Comment (jviau) [inline src/DurableTask.Core/History/TimerCreatedEvent.cs:51]: Does `ParentTraceContext` need to be stored? I am wondering if `Activity.Current` will already be re-hydrated to the appropriate parent when timer event span is to be emitted. And as for storing this timers span itself, I don't think it needs to be stored at all? We only need to create the `Acitivty` and emit it right when the timer completes. (I think at least).

File: src/DurableTask.Core/TaskOrchestrationContext.cs [hunk 4]
```diff
@@ -484,6 +484,12 @@ public void HandleTimerFiredEvent(TimerFiredEvent timerFiredEvent)
             {
                 LogDuplicateEvent("TimerFired", timerFiredEvent, taskId);
             }
+
+            if (!this.IsReplaying)
+            {
+                Activity activity = TraceHelper.CreateActivityForTimer(this.OrchestrationInstance, timerFiredEvent.Timestamp, timerFiredEvent.FireAt);
```
Comment (jviau) [inline src/DurableTask.Core/TaskOrchestrationContext.cs:490]: If you add `using` here you can remove the `activity.Stop()`. The disposal will call `Stop()` for you. Plus, `activity` might be `null` - the using will handle that case for you.

```suggestion
                using Activity activity = TraceHelper.CreateActivityForTimer(this.OrchestrationInstance, timerFiredEvent.Timestamp, timerFiredEvent.FireAt);
```

File: src/DurableTask.Core/History/TimerCreatedEvent.cs [hunk 5]
```diff
@@ -41,5 +43,39 @@ public TimerCreatedEvent(int eventId)
...
+            }
+
+            return ActivityContext.TryParse(
+                this.ParentTraceContext.TraceParent,
+                this.ParentTraceContext.TraceState,
+                out parentTraceContext);
+        }
+
+        // Used for new orchestration and sub-orchestration
+        internal void SetParentTraceContext(Activity traceActivity)
```
Comment (jviau) [inline src/DurableTask.Core/History/TimerCreatedEvent.cs:71]: This code is duplicated in many history events. Can we consolidate it a bit?

1. First, make sure it is absolutely necessary. Can some of these events rely on `Activity.Current` being available instead of storing them? There is a cost to storing this extra property on these events, lets only pay it if we absolutely have to.
2. You could make an interface and extension methods:

``` CSharp
internal interface IDurableContextHolder
{
    DistributedTraceContext TraceContext { get; set; }
}

internal static DistributedContextHolderExtensions
{
    public static bool TryGetParentTraceContext(this IDurableContextHolder holder, out ActivityContext parentContext)
    {
        // implementation
    }

    public static void SetParentTraceContext(this IDurableContextHolder holder, Activity activity)
    {
        // implementation
    }
}
```

**Comment by bachuv** [comment]
@jviau do you have any comments/feedback about the responses that @cgillum and I gave regarding the different timer scenarios?

File: src/DurableTask.Core/TaskOrchestrationDispatcher.cs [hunk 6]
```diff
@@ -697,6 +697,11 @@ bool ReconcileMessagesWithState(TaskOrchestrationWorkItem workItem)
                     continue;
                 }
 
+                if (Activity.Current != null && message.Event is TimerFiredEvent)
```
Comment (cgillum) [inline src/DurableTask.Core/TaskOrchestrationDispatcher.cs:700]: You can avoid the cast on line 702 by declaring a variable in your `is` expression.
```suggestion
                if (Activity.Current != null && message.Event is TimerFiredEvent timerFired)
```

File: src/DurableTask.Core/TaskOrchestrationDispatcher.cs [hunk 7]
```diff
@@ -697,6 +697,11 @@ bool ReconcileMessagesWithState(TaskOrchestrationWorkItem workItem)
...
+                {
+                    using Activity activity = TraceHelper.CreateActivityForTimer(workItem.OrchestrationRuntimeState.OrchestrationInstance, message.Event.Timestamp, ((TimerFiredEvent)message.Event).FireAt);
```
Comment (cgillum) [inline src/DurableTask.Core/TaskOrchestrationDispatcher.cs:702]: nit: Now that I'm looking at this, it feels like a very unintuitive use of the C# `using` feature. I think it would be easier for readers of the to understand what we're doing if we disposed the activity explicitly, instead of implicitly. That also reduces the chance that someone accidentally introduces a bug by moving curly braces around. I also suggest a comment explaining what we're doing here since it's a bit counterintuitive.
```suggestion
                    // We immediately publish the activity span for this timer by creating the activity and immediately calling Dispose() on it.
                    TraceHelper.CreateActivityForTimer(workItem.OrchestrationRuntimeState.OrchestrationInstance, message.Event.Timestamp, ((TimerFiredEvent)message.Event).FireAt).Dispose();
```

File: src/DurableTask.Core/IDurableTraceContextWrapper.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+using System.Diagnostics;
+using System.Text;
+using DurableTask.Core.Tracing;
+
+namespace DurableTask.Core
+{
+    /// <summary>
+    /// Interface to store the Trace Context for events. Used for Distributed Tracing.
+    /// </summary>
+    public interface IDurableTraceContextWrapper
```
Comment (cgillum) [inline src/DurableTask.Core/IDurableTraceContextWrapper.cs:25]: I was initially wondering if this could be internal instead of public, but I'm guessing we need this to be public so that backends (which are in separate projects) can access the trace context...

Unrelated, but I don't really like the name "Wrapper" since we're not actually wrapping anything. Another suggestion would be `ISupportsDurableTraceContext`.

**Review by cgillum — COMMENTED** [review]
A few small comments.

**Review by cgillum — APPROVED** [review]
LGTM!

File: src/DurableTask.Core/Tracing/TraceHelper.cs [hunk 9]
```diff
@@ -165,6 +165,24 @@ static TraceHelper()
...
         }
 
+        internal static Activity? CreateActivityForTimer(OrchestrationInstance instance, DateTime startTime, DateTime fireAt)
+        {
+            Activity? newActivity = ActivityTraceSource.StartActivity(
+                name: "Timer",
+                kind: ActivityKind.Internal,
+                startTime: startTime,
+                parentContext: Activity.Current?.Context ?? default,
+                tags: new KeyValuePair<string, object?>[]
```
Comment (jviau) [inline src/DurableTask.Core/Tracing/TraceHelper.cs:175]: Small suggestion (I think Rohit first mentioned this): We can have avoid allocating this array if you add the tags after activity creation and only if it is non-null.

``` CSharp
 Activity? newActivity = ActivityTraceSource.StartActivity(...);

if (newActivity is not null)
{
    newActivity.AddTag(...);
    newActivity.AddTag(...);
    newActivity.AddTag(...);
}

return newActivity;
```

**Review by jviau — APPROVED** [review]
